tests/clone_test.go

```diff
@@ -874,3 +729,149 @@ func createSourceVM(options ...libvmi.Option) (*virtv1.VirtualMachine, error) {
+	var contentsName string
+	var snapshot *snapshotv1.VirtualMachineSnapshot
+	EventuallyWithOffset(1, func() error {
+		var err error
+		snapshot, err = kubevirt.Client().VirtualMachineSnapshot(snapshotNamespace).Get(context.Background(), snapshotName, v1.GetOptions{})
+		ExpectWithOffset(2, err).ToNot(HaveOccurred())
+		if snapshot.Status == nil {
+			return fmt.Errorf("snapshot's status is nil")
+		}
+
+		if snapshot.Status.VirtualMachineSnapshotContentName != nil {
+			contentsName = *snapshot.Status.VirtualMachineSnapshotContentName
+		} else {
+			return fmt.Errorf("vm snapshot contents name is nil")
+		}
+
+		return nil
+	}, 30*time.Second, 1*time.Second).ShouldNot(HaveOccurred())
+
+	EventuallyWithOffset(1, func() error {
+		_, err := kubevirt.Client().VirtualMachineSnapshotContent(snapshotNamespace).Get(context.Background(), contentsName, v1.GetOptions{})
+		return err
+	}).ShouldNot(HaveOccurred())
+
+	return snapshot
+}
+
+func generateCloneFromVMWithParams(sourceVMName, sourceVMNamespace, targetVMName string) *clone.VirtualMachineClone {
+	vmClone := kubecli.NewMinimalCloneWithNS("testclone", sourceVMNamespace)
+
+	cloneSourceRef := &k8sv1.TypedLocalObjectReference{
+		APIGroup: pointer.P(vmAPIGroup),
+		Kind:     "VirtualMachine",
+		Name:     sourceVMName,
+	}
+
+	cloneTargetRef := cloneSourceRef.DeepCopy()
+	cloneTargetRef.Name = targetVMName
+
+	vmClone.Spec.Source = cloneSourceRef
+	vmClone.Spec.Target = cloneTargetRef
+
+	return vmClone
+}
+
+func generateCloneFromSnapshot(snapshotName, snapshotNamespace, targetVMName string) *clone.VirtualMachineClone {
+	vmClone := kubecli.NewMinimalCloneWithNS("testclone", snapshotNamespace)
+
+	cloneSourceRef := &k8sv1.TypedLocalObjectReference{
+		APIGroup: pointer.P(virtsnapshot.GroupName),
+		Kind:     "VirtualMachineSnapshot",
+		Name:     snapshotName,
+	}
+
+	cloneTargetRef := &k8sv1.TypedLocalObjectReference{
+		APIGroup: pointer.P(vmAPIGroup),
+		Kind:     "VirtualMachine",
+		Name:     targetVMName,
+	}
+
+	vmClone.Spec.Source = cloneSourceRef
+	vmClone.Spec.Target = cloneTargetRef
+
+	return vmClone
+}
+
+func createCloneAndWaitForFinish(vmClone *clone.VirtualMachineClone) {
+	By(fmt.Sprintf("Creating clone object %s", vmClone.Name))
+	vmClone, err := kubevirt.Client().VirtualMachineClone(vmClone.Namespace).Create(context.Background(), vmClone, v1.CreateOptions{})
+	Expect(err).ShouldNot(HaveOccurred())
+
+	By(fmt.Sprintf("Waiting for the clone %s to finish", vmClone.Name))
+	Eventually(func() clone.VirtualMachineClonePhase {
+		vmClone, err := kubevirt.Client().VirtualMachineClone(vmClone.Namespace).Get(context.Background(), vmClone.Name, v1.GetOptions{})
+		Expect(err).ShouldNot(HaveOccurred())
+
+		return vmClone.Status.Phase
+	}, 3*time.Minute, 3*time.Second).Should(Equal(clone.Succeeded), "clone should finish successfully")
+}
+
+func expectVMRunnable(vm *virtv1.VirtualMachine, login console.LoginToFunction) *virtv1.VirtualMachine {
```

Comment:
Same

Reply:
This function is over bloated, I plan to tackle it in a follow up PR.